Use the Fortran 2003 intrinsics command_argument_count/get_command_argument
(via small bind(C) wrappers in a new xgenie_args.f90) to obtain the command line
in the xgenie interface, instead of guessing the Fortran runtime's IARGC/GETARG
symbol names from the C compiler's predefined macros: xgenie.h mapped every
__GNUC__-defining C compiler, including the clang-based Intel icx, to gfortran's
_gfortran_iargc/_gfortran_getarg_i8, which fail to link with ifort/ifx.
Author: Simon Pinches (ITER Organization)
diff -ruN GHOST-cba30b4.orig/src/lib/CMakeLists.txt GHOST-cba30b4/src/lib/CMakeLists.txt
--- GHOST-cba30b4.orig/src/lib/CMakeLists.txt	2021-05-14 15:41:03.000000000 +0200
+++ GHOST-cba30b4/src/lib/CMakeLists.txt	2026-09-15 20:32:27.347088000 +0200
@@ -598,6 +598,7 @@
 g4putk.c
 lenstr.f
 xgenieIF.c
+xgenie_args.f90
 Opcodes.h
 xgenie.h
 )
diff -ruN GHOST-cba30b4.orig/src/lib/xgenie_args.f90 GHOST-cba30b4/src/lib/xgenie_args.f90
--- GHOST-cba30b4.orig/src/lib/xgenie_args.f90	1970-01-01 01:00:00.000000000 +0100
+++ GHOST-cba30b4/src/lib/xgenie_args.f90	2026-09-15 20:32:27.324565499 +0200
@@ -0,0 +1,24 @@
+! Command-line access for xgenieIF.c through Fortran 2003 intrinsics, so that
+! the C code no longer has to guess the Fortran runtime's IARGC/GETARG symbols
+! (which differ between gfortran, ifort/ifx and others).
+
+function ghost_arg_count() bind(C, name='ghost_arg_count') result(n)
+  use iso_c_binding, only: c_int
+  implicit none
+  integer(c_int) :: n
+  n = command_argument_count()
+end function ghost_arg_count
+
+! Fill buf(1:buflen) with argument i, blank padded (GETARG semantics)
+subroutine ghost_get_arg(i, buf, buflen) bind(C, name='ghost_get_arg')
+  use iso_c_binding, only: c_int, c_char
+  implicit none
+  integer(c_int), value :: i, buflen
+  character(kind=c_char) :: buf(*)
+  character(len=buflen) :: arg
+  integer :: k
+  call get_command_argument(i, arg)
+  do k = 1, buflen
+    buf(k) = arg(k:k)
+  end do
+end subroutine ghost_get_arg
diff -ruN GHOST-cba30b4.orig/src/lib/xgenie.h GHOST-cba30b4/src/lib/xgenie.h
--- GHOST-cba30b4.orig/src/lib/xgenie.h	2021-05-14 15:41:03.000000000 +0200
+++ GHOST-cba30b4/src/lib/xgenie.h	2026-09-15 20:32:27.346733237 +0200
@@ -70,14 +70,6 @@
 #  define STDCALL
 #endif
 
-#ifndef __INTEL_COMPILER
-#ifdef __GNUC__
-#  undef IARGC
-#  undef GETARG
-#  define IARGC  _gfortran_iargc
-#  define GETARG _gfortran_getarg_i8
-#endif
-#endif
 
 /****************  GLOBAL VARIABLES     ********************************/
 /****************  STATIC VARIABLES     ********************************/
diff -ruN GHOST-cba30b4.orig/src/lib/xgenieIF.c GHOST-cba30b4/src/lib/xgenieIF.c
--- GHOST-cba30b4.orig/src/lib/xgenieIF.c	2021-05-14 15:41:03.000000000 +0200
+++ GHOST-cba30b4/src/lib/xgenieIF.c	2026-09-15 20:32:27.346490000 +0200
@@ -48,12 +48,9 @@
 static char ServerPath[256];
 /****************  FUNCTION DECLARATIONS********************************/
 void STDCALL XINIT();
-int STDCALL IARGC();
-#ifdef _WIN32
-void STDCALL GETARG(int*,char*,int,int*);
-#else
-void STDCALL GETARG(int*,char*,int);
-#endif
+/* Fortran 2003 wrappers in xgenie_args.f90 */
+int  ghost_arg_count(void);
+void ghost_get_arg(int, char*, int);
 int  GetArgCount();
 void GetArg(int*, char*, int);
 void STDCALL XPOLYL(int* xpoints, int* ypoints, int* numpoints);
@@ -131,7 +128,7 @@
 
 /*
  * 0. Command line argument processing.
- *    have to call the FORTRAN functions IARGC and GETARG
+ *    via the Fortran 2003 wrappers in xgenie_args.f90
  */
 
   argc = GetArgCount();
@@ -292,11 +289,7 @@
  ***********************************************************************/
 int GetArgCount()
 {
-#ifdef hp
-  return(IARGC()-1);
-#else
-  return(IARGC());
-#endif
+  return(ghost_arg_count());
 }
 /************************************************************************
  * Function Name   :  GetArg
@@ -308,16 +301,7 @@
      char *Buffer;
      int len;
 {
-#if hp
-  int j = (*i)+1;
-  /* HP arg numbers are all incremented by 1 */
-  GETARG(&j,Buffer,len);
-#elif defined _WIN32
-  int status;
-  GETARG(i,Buffer,len,&status);
-#else
-  GETARG(i,Buffer,len);
-#endif
+  ghost_get_arg(*i,Buffer,len);
 }
 
 
